central/sensor/service/pipeline/clustermetrics/update_maxima.go

```diff
@@ -0,0 +1,41 @@
+package clustermetrics
+
+import (
+	"github.com/stackrox/rox/generated/internalapi/central"
+	"github.com/stackrox/rox/pkg/maputil"
+	"github.com/stackrox/rox/pkg/set"
+)
+
+// BillingMetrics are the metrics we collect and show to the customers to help
+// them report their usage.
+type BillingMetrics struct {
+	TotalNodes int64
+	TotalCores int64
+}
+
+var (
+	nodesMap = maputil.NewMaxMap[string, int64]()
+	coresMap = maputil.NewMaxMap[string, int64]()
+)
+
+func updateMaxima(clusterID string, cm *central.ClusterMetrics) {
+	nodesMap.Add(clusterID, cm.GetNodeCount())
+	coresMap.Add(clusterID, cm.GetCpuCapacity())
+}
+
+// CutMetrics resets the metrics and returns the collected values since last
+// invocation.
+func CutMetrics(ids set.StringSet) *BillingMetrics {
+	var m BillingMetrics
+	for id, v := range nodesMap.Reset() {
+		if ids.Contains(id) {
```

Comment:
What is the purpose of this condition? Doesn't this lead to a wrong output, because it does not calculate the max value over the last period, but the max value of the currently active clusters?

Reply:
I would propose the following algorithm to calculate the hourly max instead:

* Keep the last value of each cluster and the total max in memory.
* When a new pipeline message comes in, we update the cluster value for ^.
* Use pipeline exit function to set value to zero / remove cluster.
* We update the new max like `new_max_value = max(old_max_value, new_max_value)`.
* After 1 hour, we reset `new_max_value = sum(cluster_values)`.

This way we should handle both different clusters popping in and out, and the same cluster changing IDs correctly.